frame/contracts/src/rent.rs

```diff
@@ -35,56 +35,49 @@ use sp_runtime::{
 	traits::{Bounded, CheckedDiv, CheckedMul, SaturatedConversion, Saturating, Zero},
 };
 
-/// The amount to charge.
+/// Information about the required deposit and resulting rent.
 ///
-/// This amount respects the contract's rent allowance and the subsistence deposit.
-/// Because of that, charging the amount cannot remove the contract.
-struct OutstandingAmount<T: Config> {
-	amount: BalanceOf<T>,
+/// The easiest way to guarantee that a contract stays alive is to assert that
+/// `max_rent == 0` at the **end** of a contract's execution.
+///
+/// # Note
+///
+/// The `current_*` fields do **not** consider changes to the code's refcount made during
+/// the currently running call.
+#[derive(codec::Encode)]
+#[cfg_attr(test, derive(Debug, PartialEq))]
+pub struct RentStatus<T: Config> {
+	/// Required deposit assuming that this contract is the only user of its code.
+	pub max_deposit: BalanceOf<T>,
+	/// Required deposit assuming the code's current refcount.
+	pub current_deposit: BalanceOf<T>,
+	/// Required deposit assuming the specified refcount (None if 0 is supplied).
+	pub custom_refcount_deposit: Option<BalanceOf<T>>,
+	/// Rent that is payed assuming that the contract is the only user of its code.
+	pub max_rent: BalanceOf<T>,
+	/// Rent that is payed given the code's current refcount.
+	pub current_rent: BalanceOf<T>,
+	/// Rent that is payed assuming the specified refcount (None is 0 is supplied).
+	pub custom_refcount_rent: Option<BalanceOf<T>>,
+	/// Reserved for backwards compatible changes to this data structure.
+	pub _reserved: Option<()>,
 }
 
-impl<T: Config> OutstandingAmount<T> {
-	/// Create the new outstanding amount.
-	///
-	/// The amount should be always withdrawable and it should not kill the account.
-	fn new(amount: BalanceOf<T>) -> Self {
-		Self { amount }
-	}
-
-	/// Returns the amount this instance wraps.
-	fn peek(&self) -> BalanceOf<T> {
-		self.amount
-	}
-
-	/// Withdraws the outstanding amount from the given account.
-	fn withdraw(self, account: &T::AccountId) {
-		if let Ok(imbalance) = T::Currency::withdraw(
-			account,
-			self.amount,
-			WithdrawReasons::FEE,
-			ExistenceRequirement::KeepAlive,
-		) {
-			// This should never fail. However, let's err on the safe side.
-			T::RentPayment::on_unbalanced(imbalance);
+/// We cannot derive `Default` because `T` does not necessarily implement `Default`.
```

Comment:
Would `frame_support::DefaultNoBound` do the job?

Reply:
Nice. Did not know about this. Was able to remove some boilerplate.